### PR TITLE
BUG: Format floats using their intrinsic decimal precision

### PR DESCRIPTION
**Pull request by programmarchy**
Since `FloatObject` is represented as a decimal, format numbers using their intrinsic precision, instead of reducing the precision to 5 decimal places.

This fixes rendering issues for PDFs that contain coordinates, transformations, etc. with real numbers containing more than 5 decimal places of precision. For example, PDFs exported from Microsoft PowerPoint contain numbers with up to 11 decimal places.

Fixes: https://github.com/py-pdf/PyPDF2/issues/1266

### CODE REVIEW

File: tests/test_writer.py [hunk 1]
```diff
@@ -777,7 +777,7 @@ def test_colors_in_outline_item():
     reader2 = PdfReader(target)
     for outline_item in reader2.outline:
         # convert float to string because of mutability
-        assert [str(c) for c in outline_item.color] == [str(p) for p in purple_rgb]
+        assert [f"{c:.5f}" for c in outline_item.color] == [f"{p:.5f}" for p in purple_rgb]
```
Comment (MasterOdin) [inline tests/test_writer.py:780]: Why was this change necessary? What is getting outputted for `str(c)` and `str(p)` vs this new formatting?

File: PyPDF2/generic/_base.py [hunk 2]
```diff
@@ -246,7 +246,7 @@ def __repr__(self) -> str:
             return str(self.quantize(decimal.Decimal(1)))
         else:
             # Standard formatting adds useless extraneous zeros.
-            o = f"{self:.5f}"
+            o = '{0:f}'.format(self)
             # Remove the zeros.
             while o and o[-1] == "0":
                 o = o[:-1]
```
Comment (MasterOdin) [inline PyPDF2/generic/_base.py:252]: It looks like we could simplify this else branch down to just this one liner?
```
            return '{0:f}'.format(self).rstrip('0')
```

File: tests/test_writer.py (hunk 1 above)
Comment (programmarchy) [reply tests/test_writer.py:780]: Good question! "Hidden" in this test was the assumption that floats are represented with 5 decimal places, because they're baked into the `purple_rgb` assignment:

```
    purple_rgb = (0.50196, 0, 0.50196)
```

However, since a `FloatObject` gets turned into a decimal, the in-memory representation will become:

```
>>> Decimal(0.50196)
Decimal('0.501959999999999961772800816106610000133514404296875')
```

So this patch uses the `.5f` specifier to explicitly match the `purple_rgb` assignment.

Comment (MasterOdin) [reply tests/test_writer.py:780]: Okay, so we could keep `str(p)` and then only apply the formatting to values coming from `c`. 

If we were to set `purple_rgb == ('0.50196', 0, '0.50196')`, then would we be able to keep the existing test asserts as-is?

Though, this is surprising to me, as it looks like we should always be casting the value to a string _before_ we convert to a decimal, so we should be doing:
```
>>> a = Decimal('0.50196')
>>> a
Decimal('0.50196')
>>> str(a)
'0.50196'
>>> '{0:f}'.format(a)
'0.50196'

File: PyPDF2/generic/_base.py (hunk 2 above)
Comment (programmarchy) [reply PyPDF2/generic/_base.py:252]: Hmm, I on second thought there may be an edge case here, where we don't want to strip the last zero.

The while loop suffers the same problem as `rstrip` though, but shouldn't "0.000" become "0.0" instead of "0."?

Comment (MasterOdin) [reply PyPDF2/generic/_base.py:252]: Oh, true.

something like:
```
o = '{0:f}'.format(self).rstrip('0')
if o[-1] == '.':
  o += "0"
return o
```

Comment (MasterOdin) [reply PyPDF2/generic/_base.py:252]: Hm, though this also supposes that we'll always have a decimal number and not something like `Decimal('100')`

Comment (MasterOdin) [reply PyPDF2/generic/_base.py:252]: Maybe (without testing):
```
o = '{0:f}'.format(self)
if o and '.' in o:
  o = o.rstrip('0')
  if o and o[-1] == '.':
    o += "0"
return o
```

File: tests/test_writer.py (hunk 1 above)
Comment (programmarchy) [reply tests/test_writer.py:780]: > Okay, so we could keep `str(p)` and then only apply the formatting to values coming from `c`.

Almost, but not quite, because the second value in that tuple (zero) does not have 5 decimal places:

```
E           AssertionError: assert ['0.50196', '0.00000', '0.50196'] == ['0.50196', '0', '0.50196']
E             At index 1 diff: '0.00000' != '0'
E             Use -v to get more diff
```

We'd be violating the type constraint if we used strings in the tuple:

```
color: Optional[Union[Tuple[float, float, float], str]] = None,
```

I think I see where you're going using strings, though. The test still fails with that change, so need to do a little more digging to find out why.

File: PyPDF2/generic/_base.py (hunk 2 above)
Comment (programmarchy) [reply PyPDF2/generic/_base.py:252]: Ah, yep. Current implementation is bugged if the input is `Decimal("0")`!

Surprising that no tests caught that... I'll add a test that does.

Comment (programmarchy) [reply PyPDF2/generic/_base.py]: Dug into this one, and we're safe to simplify to `rstrip("0")`. The reason for that is because the first condition:

```python
        if self == self.to_integral():
            # If this is an integer, format it with no decimal place.
            return str(self.quantize(decimal.Decimal(1)))
```

I added the comment to help explain what the `to_integral` and `quantize` functions from `Decimal` are doing.

If the number is an integer, that covers where there's no decimal point or there's only zeros after the decimal e.g. `1000`, `1`, `1.0`, `1.0000`, etc

I've added a unit test `test_float_object_decimal_to_string` to test common and extreme cases.

File: tests/test_writer.py (hunk 1 above)
Comment (programmarchy) [reply tests/test_writer.py:780]: Dug into this one, too.

The test passes the third case if we use strings in the `purple_rgb =('0.50196', 0, '0.50196')` assignment, however, it fails the first and second cases:

```python
    writer.add_outline_item("First Outline Item", pagenum=2, color="800080") # FAIL
    writer.add_outline_item("Second Outline Item", pagenum=3, color="#800080") # FAIL
    writer.add_outline_item("Third Outline Item", pagenum=4, color=purple_rgb) # PASS
```

The reason for this is when color is pass in as a hex string, it gets converted to a tuple of floats by `hex_to_rgb`:

```python
>>> from PyPDF2.generic import hex_to_rgb, FloatObject
>>> rgb = hex_to_rgb('#800080')
>>> rgb[0]
0.5019607843137255
>>> repr(FloatObject(rgb[0]))
'0.5019607843137254832299731788225471973419189453125'
```

So this PR does create a side effect of FloatObjects produced by `hex_to_rgb` having very high precision. One way we could deal with that is to have `hex_to_rgb` change its output from a tuple of floats to a tuple of Decimals so that we can set the precision to something reasonable like 5 decimal places.

Comment (programmarchy) [reply tests/test_writer.py:780]: @MasterOdin That is a much better idea, and will be easy to include in unit tests. I'll take a stab at it.

Comment (programmarchy) [reply tests/test_writer.py:780]: Just found out something super interesting about python's Decimal:

```python
>>> import decimal
>>> decimal.getcontext()
Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])
```

So you can call `decimal.getcontext().precision` to get the current precision. By default it's 28 (on my machine).

You can also do:

```python
with decimal.localcontext() as ctx:
  ctx.precision = 5
  # Do stuff with Decimal
```

By deferring to the decimal context, we won't need an extra parameter on `FloatObject`. Going to try it...

Comment (programmarchy) [reply tests/test_writer.py]: Implemented your suggestion, which allowed us to keep the unit test for the writer the same, and will prevent the issue with floating converted from hex values having very high precision.

However, in the future, a better solution may be to do something similar to [decimal.Context](https://github.com/python/cpython/blob/main/Lib/_pydecimal.py#L435). We wouldn't need the extra argument on FloatObject anymore, and it'd give users more control over how formatting occurs.

Roughly, it could look something like:

```python
import PyPDF2
from PyPDF2 import PdfReader, PdfWriter
from PyPDF2.context import StripExtraTrailingZeros, QuantizeInteger

with PyPDF2.localcontext() as ctx:
  ctx.prec = 5 # specify precision
  ctx.flags = [StripExtraTrailingZeros, QuantizeInteger] # specify additional format flags

  reader = PdfReader("./path/to/file.pdf")
  writer = PdfWriter()

  # context settings will be used for formatting
```

Comment (programmarchy) [reply tests/test_writer.py]: Modified the implementation a bit to make it simpler.

It seemed wrong to add a precision property to `FloatObject` since it inherits from `Decimal`, which already represents precision internally. So instead of adding an extra precision property to `FloatObject`, we can just pass it a "quantized" `Decimal` e.g. `d.quantize(Decimal('1.00000'))` sets the precision of `d` to 5.

If you agree, happy to rebase and squash the latest commit.

**Comment by programmarchy** [comment]
Rebased this PR so tests are passing, and believe all the changes requested in the last review have been addressed.

Could you take another look, @MasterOdin?

**Review by MartinThoma — APPROVED** [review]
It looks good from my side - thank you for writing a unit test :hugs:

**Comment by MartinThoma** [comment]
@MasterOdin You were a lot more involved in this PR than I was. What do you think?

**Review by MasterOdin — APPROVED** [review]

**Comment by MasterOdin** [comment]
Looks good. Thanks for the work here @programmarchy! :+1:

**Comment by MartinThoma** [comment]
Thank you both for all the work you put in it :pray: 

I've just merged the PR and I will release it today to PyPI :-)

**Comment by MartinThoma** [comment]
@programmarchy If you want, I can add you to https://pypdf2.readthedocs.io/en/latest/meta/CONTRIBUTORS.html :-)

**Comment by programmarchy** [comment]
> @programmarchy If you want, I can add you to https://pypdf2.readthedocs.io/en/latest/meta/CONTRIBUTORS.html :-)

That would be very cool, thank you @MartinThoma!

**Comment by mrknwk** [comment]
When you use `page.scale_by()` with a non-integer value in 2.10.9, Acrobat (22.002.20212) displays the transformed PDF pages as empty square drawing areas. Maybe one could leave the default precision at less than 20 digits (this is the Acrobat tipping point, I guess) for compatibility reasons and make it configurable?

**Comment by programmarchy** [comment]
Interesting, would you mind sharing how you came to find out 20 digits is the tipping point for Acrobat, @mrknwk?

One way I was thinking of to make this configurable would be to adopt [context vars](https://docs.python.org/3/library/contextvars.html) as implemented in [decimal.Context](https://github.com/python/cpython/blob/main/Lib/_pydecimal.py#L435) for example. The context provides sane defaults with a central point for changing behavior.

It would allow us to write something like:

```
import PyPDF2
from PyPDF2 import PdfReader, PdfWriter
from PyPDF2.context import Context, StripExtraTrailingZeros, QuantizeInteger

ctx = StreamContext()
ctx.max_prec = 5 # specify maximum precision
ctx.flags = [StripExtraTrailingZeros, QuantizeInteger] # could also specify additional format flags
PyPDF2.setcontext(ctx)

reader = PdfReader("./path/to/file.pdf")
reader.pages[0].scale_by(0.5)
writer = PdfWriter()
writer.add_page(reader.pages[0])
...
```

Or like this:
```
with PyPDF2.localcontext() as ctx:
  ctx.max_prec = 5 # specify maximum precision
  ctx.flags = [StripExtraTrailingZeros, QuantizeInteger] # could also specify additional format flags
  ...
```

Or maybe this:
```
PyPDF2.setcontext(AdobeAcrobactContext)
```

Might make sense to open a separate issue to discuss further.

**Comment by mrknwk** [comment]
@programmarchy It really was just trial and error. 😊 

But 20 digits is also the limit that one of the maintainers of PDF Arranger found [in a test](https://github.com/pdfarranger/pdfarranger/issues/198#issuecomment-644954633). He contacted Adobe about it and apparently it is an Acrobat "implementation level limitation". So maybe the third option would be a nice way to go then.

**Comment by programmarchy** [comment]
@mrknwk I'd be happy to take a stab at implementing the above. Could you please create a GitHub issue with a corresponding sample PDF, and tag me?
